File: circuit/withdraw.go
// Package circuit — WithdrawCircuit proves knowledge of a secret key and
// sufficient balance to withdraw a publicly known amount of SOL lamports.
//
// # Private witnesses
//
//	Sk         – account secret key (BN254 Fr)
//	RNew       – randomness for the new (post-withdraw) ciphertext
//	OldBalance – plaintext balance before withdrawal (u32)
//	NewBalance – OldBalance − Amount; proved non-negative via range check (u32)
//
// # Public inputs (5 values)
//
//	Pk              – account public key (G1)
//	OldC1, OldC2   – current on-chain ciphertext (G1, G1)
//	NewC1, NewC2   – replacement ciphertext after withdraw (G1, G1)
//	Amount          – lamports being withdrawn (u32, native frontend.Variable)
//
// The circuit proves:
//
//  1. Pk     = Sk * G                           (key ownership)
//  2. OldC2  = Sk * OldC1 + OldBalance * G      (correct decryption)
//  3. NewC1  = RNew * G                          (fresh randomness)
//  4. NewC2  = RNew * Pk + NewBalance * G        (correct re-encryption)
//  5. OldBalance = Amount + NewBalance           (arithmetic soundness)
//  6. OldBalance, NewBalance, Amount ∈ [0, 2³²) (no overdraft via range check)
package circuit

import (
	"github.com/consensys/gnark-crypto/ecc/bn254"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/std/algebra"
	"github.com/consensys/gnark/std/algebra/emulated/sw_bn254"
	"github.com/consensys/gnark/std/math/emulated"
)

// WithdrawCircuit is the Groth16 withdraw constraint system.
type WithdrawCircuit struct {
	// ── private witnesses ──────────────────────────────────────────────────
	Sk         emulated.Element[sw_bn254.ScalarField] `gnark:",secret"`
	RNew       emulated.Element[sw_bn254.ScalarField] `gnark:",secret"`
	OldBalance frontend.Variable                      `gnark:",secret"`
	NewBalance frontend.Variable                      `gnark:",secret"` // OldBalance − Amount

	// ── public inputs ──────────────────────────────────────────────────────
	Pk     sw_bn254.G1Affine `gnark:",public"` // account public key
	OldC1  sw_bn254.G1Affine `gnark:",public"` // current ciphertext C1
	OldC2  sw_bn254.G1Affine `gnark:",public"` // current ciphertext C2
	NewC1  sw_bn254.G1Affine `gnark:",public"` // replacement ciphertext C1
	NewC2  sw_bn254.G1Affine `gnark:",public"` // replacement ciphertext C2
	Amount frontend.Variable `gnark:",public"` // lamports withdrawn
}

func (c *WithdrawCircuit) Define(api frontend.API) error {
	curve, err := algebra.GetCurve[sw_bn254.ScalarField, sw_bn254.G1Affine](api)
	if err != nil {
		return err
	}
	scalarField, err := emulated.NewField[sw_bn254.ScalarField](api)
	if err != nil {
		return err
	}

	// ── 1. Range checks ────────────────────────────────────────────────────
	oldBits := api.ToBinary(c.OldBalance, 32)
	api.ToBinary(c.NewBalance, 32) // NewBalance range check
	api.ToBinary(c.Amount, 32)     // Amount range check

	// ── 2. OldBalance = Amount + NewBalance ────────────────────────────────
	api.AssertIsEqual(c.OldBalance, api.Add(c.Amount, c.NewBalance))

	// ── 3. Lift OldBalance to emulated scalar for EC ops ───────────────────
	oldScalar := scalarField.FromBits(oldBits...)

	// Generator constant used below.
	_, _, g1gen, _ := bn254.Generators()
	gPoint := sw_bn254.NewG1Affine(g1gen)

	// ── 4. Pk = Sk * G ─────────────────────────────────────────────────────
	computedPk := curve.ScalarMulBase(&c.Sk)
	curve.AssertIsEqual(computedPk, &c.Pk)

	// ── 5. OldC2 = Sk * OldC1 + OldBalance * G ────────────────────────────
	skC1 := curve.ScalarMul(&c.OldC1, &c.Sk)
	oldBalG := curve.ScalarMulBase(oldScalar)
	computedOldC2 := curve.Add(skC1, oldBalG)
	curve.AssertIsEqual(computedOldC2, &c.OldC2)

	// ── 6. NewC1 = RNew * G ────────────────────────────────────────────────
	computedNewC1 := curve.ScalarMulBase(&c.RNew)
	curve.AssertIsEqual(computedNewC1, &c.NewC1)

	// ── 7. NewC2 = RNew * Pk + NewBalance * G ─────────────────────────────
	// Compute NewBalance*G as (NewBalance+1)*G − G to avoid ScalarMulBase(0)
	// when NewBalance = 0 (full withdrawal).  Scalar NewBalance+1 ∈ [1, 2³²]
	// is never zero.  Both curve.Add calls operate on non-identity,
	// non-equal points with overwhelming probability for random RNew.
	nbP1Bits := api.ToBinary(api.Add(c.NewBalance, 1), 33)
	nbP1Scalar := scalarField.FromBits(nbP1Bits...)
	nbP1G := curve.ScalarMulBase(nbP1Scalar)
	rnPk := curve.ScalarMul(&c.Pk, &c.RNew)
	computedNewC2 := curve.Add(curve.Add(rnPk, nbP1G), curve.Neg(&gPoint))
	curve.AssertIsEqual(computedNewC2, &c.NewC2)

	return nil
}

File: circuit/src/withdraw.rs
//! Withdraw circuit.
//!
//! Proves knowledge of a secret key and sufficient balance to withdraw a
//! publicly known amount of SOL lamports.
//!
//! # Private witnesses
//!
//! - `sk`          — account BJJ secret key
//! - `r_new`       — randomness for the post-withdraw ciphertext
//! - `old_balance` — plaintext balance before withdrawal (u32)
//! - `new_balance` — `old_balance − amount` (u32); range-checked ≥ 0
//!
//! # Public inputs (11 native Fr elements)
//!
//! - `pk`     — account BJJ public key       (2 Fr)
//! - `old_c1` — current ciphertext C1        (2 Fr)
//! - `old_c2` — current ciphertext C2        (2 Fr)
//! - `new_c1` — replacement ciphertext C1    (2 Fr)
//! - `new_c2` — replacement ciphertext C2    (2 Fr)
//! - `amount` — lamports withdrawn           (1 Fr, native u32)
//!
//! Total: 11 Fr public inputs, IC length = 12.
//!
//! # Constraints
//!
//! 1. `pk     = Sk * G`
//! 2. `old_c2 = Sk * old_c1 + OldBalance * G`
//! 3. `new_c1 = RNew * G`
//! 4. `new_c2 = RNew * Pk + NewBalance * G`
//! 5. `old_balance = amount + new_balance`
//! 6. `old_balance, new_balance, amount ∈ [0, 2³²)`

use ark_bn254::Fr;
use ark_ed_on_bn254::{constraints::EdwardsVar, EdwardsAffine, Fr as BJJFr};
use ark_r1cs_std::{fields::fp::FpVar, prelude::*};
use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError};

use crate::gadgets::{
    alloc_bjj_scalar_bits, alloc_u32_bits, bits_to_fpvar, generator_var, scalar_mul,
};

#[derive(Clone)]
pub struct WithdrawCircuit {
    // private
    pub sk: Option<BJJFr>,
    pub r_new: Option<BJJFr>,
    pub old_balance: Option<u32>,
    pub new_balance: Option<u32>,

    // public
    pub pk: Option<EdwardsAffine>,
    pub old_c1: Option<EdwardsAffine>,
    pub old_c2: Option<EdwardsAffine>,
    pub new_c1: Option<EdwardsAffine>,
    pub new_c2: Option<EdwardsAffine>,
    pub amount: Option<u32>,
}

impl ConstraintSynthesizer<Fr> for WithdrawCircuit {
    fn generate_constraints(self, cs: ConstraintSystemRef<Fr>) -> Result<(), SynthesisError> {
        // ── Public inputs ──────────────────────────────────────────────────────
        let pk_var = EdwardsVar::new_input(cs.clone(), || {
            self.pk.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let old_c1_var = EdwardsVar::new_input(cs.clone(), || {
            self.old_c1.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let old_c2_var = EdwardsVar::new_input(cs.clone(), || {
            self.old_c2.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let new_c1_var = EdwardsVar::new_input(cs.clone(), || {
            self.new_c1.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let new_c2_var = EdwardsVar::new_input(cs.clone(), || {
            self.new_c2.ok_or(SynthesisError::AssignmentMissing)
        })?;
        let amount_fp = FpVar::new_input(cs.clone(), || {
            self.amount
                .map(|a| Fr::from(a as u64))
                .ok_or(SynthesisError::AssignmentMissing)
        })?;

        // ── Private witnesses ──────────────────────────────────────────────────
        let sk_bits = alloc_bjj_scalar_bits(cs.clone(), self.sk)?;
        let r_new_bits = alloc_bjj_scalar_bits(cs.clone(), self.r_new)?;

        let old_bal_bits = alloc_u32_bits(cs.clone(), self.old_balance)?;
        let new_bal_bits = alloc_u32_bits(cs.clone(), self.new_balance)?;
        let amount_bits = alloc_u32_bits(cs.clone(), self.amount)?;

        let old_bal_fp = bits_to_fpvar(&old_bal_bits);
        let new_bal_fp = bits_to_fpvar(&new_bal_bits);
        let amount_fp_reconstructed = bits_to_fpvar(&amount_bits);

        // Verify public amount matches private bit decomposition
        amount_fp.enforce_equal(&amount_fp_reconstructed)?;

        // ── Constraint 5: old_balance = amount + new_balance ──────────────────
        old_bal_fp.enforce_equal(&(amount_fp_reconstructed + &new_bal_fp))?;

        // ── Constraint 1: pk = Sk * G ──────────────────────────────────────────
        let g = generator_var();
        let computed_pk = scalar_mul(&g, &sk_bits)?;
        computed_pk.enforce_equal(&pk_var)?;

        // ── Constraint 2: old_c2 = Sk * old_c1 + OldBalance * G ───────────────
        let sk_c1 = scalar_mul(&old_c1_var, &sk_bits)?;
        let old_bal_g = scalar_mul(&g, &old_bal_bits)?;
        let computed_old_c2 = sk_c1 + old_bal_g;
        computed_old_c2.enforce_equal(&old_c2_var)?;

        // ── Constraint 3: new_c1 = RNew * G ───────────────────────────────────
        let computed_new_c1 = scalar_mul(&g, &r_new_bits)?;
        computed_new_c1.enforce_equal(&new_c1_var)?;

        // ── Constraint 4: new_c2 = RNew * Pk + NewBalance * G ─────────────────
        let rn_pk = scalar_mul(&pk_var, &r_new_bits)?;
        let new_bal_g = scalar_mul(&g, &new_bal_bits)?;
        let computed_new_c2 = rn_pk + new_bal_g;
        computed_new_c2.enforce_equal(&new_c2_var)?;

        Ok(())
    }
}
